src/mscorlib/src/System/Globalization/DateTimeFormat.cs

```diff
@@ -957,12 +949,88 @@ internal static String Format(DateTime dateTime, String format, DateTimeFormatIn
             }
 
             if (format.Length == 1) {
+                switch (format[0]) {
+                    case 'o':
+                    case 'O':
+                        // Fast track for round trip format without going through a format string.
+                        return RoundTripFormat(ref dateTime);
+                }
+
                 format = ExpandPredefinedFormat(format, ref dateTime, ref dtfi, ref offset);
-            }            
+            }      
 
             return (FormatCustomized(dateTime, format, dtfi, offset));
         }
-    
+
+        internal static string RoundTripFormat(ref DateTime dateTime)
+        {
+            StringBuilder result = StringBuilderCache.Acquire();
+            Append(result, dateTime.Year, 4);
+            result.Append('-');
+
+            Append(result, dateTime.Month, 2);
+            result.Append('-');
+
+            Append(result, dateTime.Day, 2);
+            result.Append('T');
+
+            Append(result, dateTime.Hour, 2);
+            result.Append(':');
+
+            Append(result, dateTime.Minute, 2);
+            result.Append(':');
+
+            Append(result, dateTime.Second, 2);
+            result.Append('.');
+
+            long fraction = dateTime.Ticks % TimeSpan.TicksPerSecond;
+
+            Append(result, fraction, 7);
+
+            switch (dateTime.Kind)
+            {
+                case DateTimeKind.Local:
+                    {
+                        TimeSpan offset = TimeZoneInfo.Local.GetUtcOffset(dateTime);
+
+                        if (offset >= TimeSpan.Zero)
+                        {
+                            result.Append('+');
+                        }
+                        else
+                        {
+                            result.Append('-');
+                            offset = offset.Negate();
+                        }
+
+                        Append(result, offset.Hours, 2);
+                        result.Append(':');
+                        Append(result, offset.Minutes, 2);
+                    }
+                    break;
+
+                case DateTimeKind.Utc:
+                    result.Append('Z');
+                    break;
+
+                default:
+                    break;
+            }
+
+            return StringBuilderCache.GetStringAndRelease(result);
+        }
+
+        internal static void Append(StringBuilder builder, long val, int digit)
```

Comment:
>Append [](start = 29, length = 6)

nit: could you call this AppendNumber instead. it will make easier to not confuse it with the StringBuilder Append